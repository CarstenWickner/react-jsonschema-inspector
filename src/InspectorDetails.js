/* eslint-disable quote-props */
import PropTypes from "prop-types";
import React, { PureComponent } from "react";

import InspectorDetailsForm from "./InspectorDetailsForm";
import JsonSchemaPropType from "./JsonSchemaPropType";
import { getFieldValue, getPropertyParentSchemas, isDefined } from "./utils";

class InspectorDetails extends PureComponent {
    static isSelectionRequiredInParent(columnData, selectionColumnIndex, refTargets) {
        if (selectionColumnIndex < 1) {
            // no parent to define any required properties
            return false;
        }
        const parentColumn = columnData[selectionColumnIndex - 1];
        const parentSchema = parentColumn.items[parentColumn.selectedItem];
        const targetItem = columnData[selectionColumnIndex].selectedItem;
        const schemaList = getPropertyParentSchemas(parentSchema, refTargets);
        return schemaList.some(part => isDefined(part.required) && part.required.includes(targetItem));
    }

    static collectFormFields(itemSchema, refTargets, columnData, selectionColumnIndex) {
        const formFields = [];
        const addFormField = (labelText, rowValue) => {
            if (isDefined(rowValue)) {
                formFields.push({ labelText, rowValue });
            }
        };
        const getValue = fieldName => getFieldValue(itemSchema, fieldName, refTargets);

        addFormField("Title", getValue("title"));
        addFormField("Description", getValue("description"));
        addFormField("Required",
            InspectorDetails.isSelectionRequiredInParent(columnData, selectionColumnIndex, refTargets) ? "Yes" : null);
        addFormField("Type", getValue("type"));
        addFormField("Constant Value", getValue("const"));
        addFormField("Possible Values", getValue("enum"));

        const minimum = getValue("minimum");
        const exclusiveMinimum = getValue("exclusiveMinimum");
        let minValue;
        if (isDefined(minimum)) {
            // according to JSON Schema Draft 4, "exclusiveMinimum" is a boolean and used in combination with "minimum"
            // if "exclusiveMinimum" is not defined, is is treated as "false", i.e. "minimum" is inclusive by default
            minValue = exclusiveMinimum ? `${minimum} (exclusive)` : `${minimum} (inclusive)`;
        } else {
            // according to JSON Schema Draft 6, "exclusiveMinimum" is a number and can be used instead of "minimum"
            minValue = isDefined(exclusiveMinimum) ? `${exclusiveMinimum} (exclusive)` : null;
        }
        addFormField("Min Value", minValue);

        const maximum = getValue("maximum");
        const exclusiveMaximum = getValue("exclusiveMaximum");
        let maxValue;
        if (isDefined(maximum)) {
            // according to JSON Schema Draft 4, "exclusiveMaximum" is a boolean and used in combination with "maximum"
            // if "exclusiveMaximum" is not defined, is is treated as "false", i.e. "maximum" is inclusive by default
            maxValue = exclusiveMaximum ? `${maximum} (exclusive)` : `${maximum} (inclusive)`;
        } else {
            // according to JSON Schema Draft 6, "exclusiveMaximum" is a number and can be used instead of "maximum"
            maxValue = isDefined(exclusiveMaximum) ? `${exclusiveMaximum} (exclusive)` : null;
        }
        addFormField("Max Value", maxValue);

        const defaultValue = getValue("default");
        addFormField("Default Value", typeof defaultValue === "object" ? JSON.stringify(defaultValue) : defaultValue);
        let examples = getValue("examples");
        examples = (isDefined(examples) && examples.length > 0) ? examples : null;
        addFormField("Example(s)",
            (examples && typeof examples[0] === "object") ? JSON.stringify(examples) : examples);
        addFormField("Value Pattern", getValue("pattern"));
        addFormField("Value Format", getValue("format"));
        addFormField("Min Length", getValue("minLength"));
        addFormField("Max Length", getValue("maxLength"));
        addFormField("Min Items", getValue("minItems"));
        addFormField("Max Items", getValue("maxItems"));
        addFormField("Items Unique", getValue("uniqueItems") === true ? "Yes" : null);

        return formFields;
    }

    constructor(props) {
        super(props);
        this.renderDefaultSelectionDetails = this.renderDefaultSelectionDetails.bind(this);
    }

    renderDefaultSelectionDetails(parameters) {
        return (
            <div className="jsonschema-inspector-details-content">
                <h3 className="jsonschema-inspector-details-header">Details</h3>
                {this.renderDetailsForm(parameters)}
            </div>
        );
    }

    renderDetailsForm({
        itemSchema, refTargets, columnData, selectionColumnIndex
    }) {
        const formFields = InspectorDetails.collectFormFields(itemSchema, refTargets, columnData, selectionColumnIndex);

        const arrayItems = getFieldValue(itemSchema, "items", refTargets);
        // look-up the kind of value expected in the array (if the schema refers to an array)
        const arrayItemSchema = ((typeof arrayItems === "object") && arrayItems)
            || ((arrayItems === true) && getFieldValue(itemSchema, "additionalItems", refTargets));

        return (
            <div>
                <InspectorDetailsForm fields={formFields} />
                {arrayItemSchema && <hr className="jsonschema-inspector-details-separator" />}
                {arrayItemSchema && <h4 className="jsonschema-inspector-details-header">Array Entry Details</h4>}
                {arrayItemSchema && this.renderDetailsForm({
                    itemSchema: arrayItemSchema,
                    refTargets,
                    columnData,
                    selectionColumnIndex: -1
                })}
            </div>
        );
    }

    render() {
        const {
            columnData, refTargets, renderSelectionDetails, renderEmptyDetails
        } = this.props;
        const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
        const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
        const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
        const itemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
        return (
            <div className="jsonschema-inspector-details">
                {itemSchema
                    && (renderSelectionDetails || this.renderDefaultSelectionDetails)({
                        itemSchema,
                        refTargets,
                        columnData,
                        selectionColumnIndex
                    })}
                {!itemSchema
                    && renderEmptyDetails
                    && renderEmptyDetails({
                        rootColumnSchemas: columnData.length ? columnData[0].items : {}
                    })}
            </div>
        );
    }
}

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    refTargets: PropTypes.objectOf(JsonSchemaPropType).isRequired,
    /** func({ itemSchema: JsonSchema, columnData, refTargets, selectionColumnIndex: number }) */
    renderSelectionDetails: PropTypes.func,
    /** func({ rootColumnSchemas }) */
    renderEmptyDetails: PropTypes.func
};

InspectorDetails.defaultProps = {
    renderSelectionDetails: null,
    renderEmptyDetails: null
};

export default InspectorDetails;
