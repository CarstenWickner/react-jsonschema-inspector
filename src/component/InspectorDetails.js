import PropTypes from "prop-types";
import React from "react";

import InspectorDetailsContent from "./InspectorDetailsContent";

import JsonSchema from "../model/JsonSchema";

const InspectorDetails = ({
    columnData, renderSelectionDetails, renderEmptyDetails
}) => {
    const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
    const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
    const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
    const itemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
    return (
        <div className="jsonschema-inspector-details">
            {itemSchema && renderSelectionDetails && renderSelectionDetails({
                itemSchema,
                selectionColumnIndex,
                columnData
            })}
            {itemSchema && !renderSelectionDetails && (
                <InspectorDetailsContent
                    itemSchema={itemSchema}
                    selectionColumnIndex={selectionColumnIndex}
                    columnData={columnData}
                />
            )}
            {!itemSchema && renderEmptyDetails && renderEmptyDetails({
                rootColumnSchemas: columnData.length ? columnData[0].items : {}
            })}
        </div>
    );
};

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchema)).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    /** func({ itemSchema: JsonSchema, selectionColumnIndex: number, columnData }) */
    renderSelectionDetails: PropTypes.func,
    /** func({ rootColumnSchemas }) */
    renderEmptyDetails: PropTypes.func
};

InspectorDetails.defaultProps = {
    renderSelectionDetails: null,
    renderEmptyDetails: null
};

export default InspectorDetails;
