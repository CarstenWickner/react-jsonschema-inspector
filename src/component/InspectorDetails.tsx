import PropTypes from "prop-types";
import React, { Component } from "react";

import InspectorDetailsContent from "./InspectorDetailsContent";
import { getColumnDataPropTypeShape } from "./renderDataUtils";
import { RenderColumn, RenderSelectionDetailsFunction, RenderEmptyDetailsFunction, RenderItemsColumn, RenderOptionsColumn } from "../types/Inspector";
import JsonSchemaGroup from "../model/JsonSchemaGroup";

interface InspectorDetailsDefaultProps {
    renderSelectionDetails: RenderSelectionDetailsFunction,
    renderEmptyDetails: RenderEmptyDetailsFunction
};

interface InspectorDetailsProps extends InspectorDetailsDefaultProps {
    columnData: Array<RenderColumn>
};

class InspectorDetails extends Component<InspectorDetailsProps> {
    render() {
        const {
            columnData, renderSelectionDetails, renderEmptyDetails
        } = this.props;
        const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
        const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
        const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
        let itemSchemaGroup: JsonSchemaGroup;
        let optionIndexes: Array<number>;
        if (trailingSelectionColumn) {
            itemSchemaGroup = (trailingSelectionColumn as RenderItemsColumn).items
                ? (trailingSelectionColumn as RenderItemsColumn).items[trailingSelectionColumn.selectedItem as string]
                : (trailingSelectionColumn as RenderOptionsColumn).contextGroup;
            if ((trailingSelectionColumn as RenderOptionsColumn).options) {
                optionIndexes = trailingSelectionColumn.selectedItem as Array<number>;
            }
        }
        return (
            <div className="jsonschema-inspector-details">
                {itemSchemaGroup && renderSelectionDetails && renderSelectionDetails({
                    itemSchemaGroup,
                    selectionColumnIndex,
                    columnData,
                    optionIndexes
                })}
                {itemSchemaGroup && !renderSelectionDetails && (
                    <InspectorDetailsContent
                        itemSchemaGroup={itemSchemaGroup}
                        selectionColumnIndex={selectionColumnIndex}
                        columnData={columnData}
                    />
                )}
                {!itemSchemaGroup && renderEmptyDetails && renderEmptyDetails({
                    rootColumnSchemas: columnData.length ? (columnData[0] as RenderItemsColumn).items : {}
                })}
            </div>
        );
    };

    static propTypes = {
        columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(false))).isRequired,
        /** func({ itemSchema: JsonSchema, selectionColumnIndex: number, columnData }) */
        renderSelectionDetails: PropTypes.func,
        /** func({ rootColumnSchemas }) */
        renderEmptyDetails: PropTypes.func
    };

    static defaultProps: InspectorDetailsDefaultProps = {
        renderSelectionDetails: null,
        renderEmptyDetails: null
    };
}

export default InspectorDetails;
