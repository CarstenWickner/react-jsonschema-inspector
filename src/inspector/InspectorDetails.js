import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import JsonSchemaPropType from './JsonSchemaPropType';
import { getPropertyValue } from './utils';

class InspectorDetails extends PureComponent {
    constructor(props) {
        super(props);
        this.renderDefaultSelectionDetails = this.renderDefaultSelectionDetails.bind(this);
    }

    render() {
        const { columnData, renderSelectionDetails, renderEmptyDetails } = this.props;
        const trailingSelectionColumnIndex = columnData.length - (columnData[columnData.length - 1].trailingSelection ? 1 : 2);
        const trailingSelectionColumn = trailingSelectionColumnIndex < 0 ? null : columnData[trailingSelectionColumnIndex];
        const selectedItemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
        return (
            <div className="jsonschema-inspector-details">
                {selectedItemSchema && (renderSelectionDetails || this.renderDefaultSelectionDetails)(selectedItemSchema, columnData, trailingSelectionColumnIndex)}
                {!selectedItemSchema && renderEmptyDetails && renderEmptyDetails(columnData)}
            </div>
        );
    }

    renderDefaultSelectionDetails(itemSchema, columnData, selectionColumnIndex) {
        return (
            <div className="jsonschema-inspector-details-content">
                <h3 className="jsonschema-inspector-details-header">Details</h3>
                {this.renderDetailsForm(itemSchema, columnData, selectionColumnIndex)}
            </div>
        );
    }

    renderDetailsForm(itemSchema, columnData, selectionColumnIndex) {
        const { refTargets } = columnData[0];
        const isRequired = this.isSelectionRequiredInParent(columnData, selectionColumnIndex);

        const title = getPropertyValue(itemSchema, 'title', refTargets);
        const description = getPropertyValue(itemSchema, 'description', refTargets);
        const type = getPropertyValue(itemSchema, 'type', refTargets);
        const constValue = getPropertyValue(itemSchema, 'const', refTargets);
        const enumValue = getPropertyValue(itemSchema, 'enum', refTargets);
        const defaultValue = getPropertyValue(itemSchema, 'default', refTargets);
        const examples = getPropertyValue(itemSchema, 'examples', refTargets);
        const pattern = getPropertyValue(itemSchema, 'pattern', refTargets);
        const format = getPropertyValue(itemSchema, 'format', refTargets);
        const minimum = getPropertyValue(itemSchema, 'minimum', refTargets);
        const exclusiveMinimum = getPropertyValue(itemSchema, 'exclusiveMinimum', refTargets);
        const maximum = getPropertyValue(itemSchema, 'maximum', refTargets);
        const exclusiveMaximum = getPropertyValue(itemSchema, 'exclusiveMaximum', refTargets);
        const minLength = getPropertyValue(itemSchema, 'minLength', refTargets);
        const maxLength = getPropertyValue(itemSchema, 'maxLength', refTargets);
        const minItems = getPropertyValue(itemSchema, 'minItems', refTargets);
        const maxItems = getPropertyValue(itemSchema, 'maxItems', refTargets);
        const uniqueItems = getPropertyValue(itemSchema, 'uniqueItems', refTargets);

        const arrayItems = getPropertyValue(itemSchema, 'items', refTargets);
        // look-up the kind of value expected in the array (if the schema refers to an array)
        const arrayItemSchema = (typeof arrayItems === 'object' && arrayItems) || (typeof arrayItems === 'boolean' && getPropertyValue(itemSchema, 'additionalItems', refTargets));

        return (
            <div>
                <form className="jsonschema-inspector-details-form">
                    {title && this.renderDetailsFormRow("Title", title)}
                    {description && this.renderDetailsFormRow("Description", description)}
                    {isRequired && this.renderDetailsFormRow("Required", "Yes")}
                    {type && this.renderDetailsFormRow("Type", type)}
                    {constValue && this.renderDetailsFormRow("Constant Value", constValue)}
                    {enumValue && this.renderDetailsFormRow("Possible Values", enumValue)}
                    {(defaultValue || defaultValue === false) && this.renderDetailsFormRow("Default Value", typeof defaultValue === 'object' ? JSON.stringify(defaultValue) : defaultValue)}
                    {examples && examples.length && this.renderDetailsFormRow("Example(s)", typeof examples[0] === 'object' ? JSON.stringify(examples) : examples)}
                    {pattern && this.renderDetailsFormRow("Value Pattern", pattern)}
                    {format && this.renderDetailsFormRow("Value Format", format)}
                    {(minimum || minimum === 0) && this.renderDetailsFormRow("Min Value", minimum + (exclusiveMinimum ? " (exclusive)" : " (inclusive)"))}
                    {!minimum && minimum !== 0 && (exclusiveMinimum || exclusiveMinimum === 0) && this.renderDetailsFormRow("Min Value", exclusiveMinimum + " (exclusive)")}
                    {(maximum || maximum === 0) && this.renderDetailsFormRow("Max Value", maximum + (exclusiveMaximum ? " (exclusive)" : " (inclusive)"))}
                    {!maximum && maximum !== 0 && (exclusiveMaximum || exclusiveMaximum === 0) && this.renderDetailsFormRow("Max Value", exclusiveMaximum + " (exclusive)")}
                    {minLength && this.renderDetailsFormRow("Min Length", minLength)}
                    {maxLength && this.renderDetailsFormRow("Max Length", maxLength)}
                    {minItems && this.renderDetailsFormRow("Min Items", minItems)}
                    {maxItems && this.renderDetailsFormRow("Max Items", maxItems)}
                    {typeof uniqueItems === 'boolean' && this.renderDetailsFormRow("Items Unique", uniqueItems ? "Yes" : "No")}
                </form>
                {arrayItemSchema && <hr />}
                {arrayItemSchema && <h4 className="jsonschema-inspector-details-header">Array Entry Details</h4>}
                {arrayItemSchema && this.renderDetailsForm(arrayItemSchema, columnData, -1)}
            </div>
        );
    }

    renderDetailsFormRow(labelText, rowValue) {
        return (
            <div className="jsonschema-inspector-details-form-row" key={labelText}>
                <label className="jsonschema-inspector-details-form-label">{labelText}:</label>
                <span className="jsonschema-inspector-details-form-value">{rowValue.toString()}</span>
            </div>
        );
    }

    isSelectionRequiredInParent(columnData, selectionColumnIndex) {
        if (selectionColumnIndex < 1) {
            // no parent to define any required properties
            return false;
        }
        const { refTargets } = columnData[0];
        const parentColumn = columnData[selectionColumnIndex - 1];
        const parentSchema = parentColumn.items[parentColumn.selectedItem];
        const requiredPropertiesInParent = getPropertyValue(parentSchema, 'required', refTargets) || [];
        const selectionName = columnData[selectionColumnIndex].selectedItem;
        return requiredPropertiesInParent.includes(selectionName);
    }
}

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        refTargets: PropTypes.objectOf(JsonSchemaPropType)
    })).isRequired,
    renderSelectionDetails: PropTypes.func, // func(selectedItemSchema: JsonSchema, columnData, selectionColumnIndex: number)
    renderEmptyDetails: PropTypes.func // func(columnData)
};

export default InspectorDetails;
