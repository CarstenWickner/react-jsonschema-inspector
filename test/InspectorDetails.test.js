import React from "react";
import { shallow } from "enzyme";
import InspectorDetails from "../src/InspectorDetails";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[
                    {
                        items: {
                            "Schema One": {
                                title: "Schema Title",
                                description: "Text"
                            }
                        },
                        selectedItem: "Schema One",
                        trailingSelection: true
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("with empty columnData", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[]}
                refTargets={{}}
            />
        );
        expect(component.children().exists()).toBe(false);
    });
    it("with empty columnData and custom renderEmptyDetails", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[]}
                refTargets={{}}
                renderEmptyDetails={({ rootColumnSchemas }) => (
                    <span className="custom-empty-details">{Object.keys(rootColumnSchemas).length}</span>
                )}
            />
        );
        expect(component.find(".custom-empty-details").text()).toBe("0");
    });
    it("with no selection", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[
                    {
                        items: {
                            "Schema One": {
                                title: "Schema Title",
                                description: "Text"
                            }
                        }
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component.children().exists()).toBe(false);
    });
    it("with no selection and custom renderEmptyDetails", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[
                    {
                        items: {
                            "Schema One": {
                                title: "Schema Title",
                                description: "Text"
                            }
                        }
                    }
                ]}
                refTargets={{}}
                renderEmptyDetails={({ rootColumnSchemas }) => (
                    <span className="custom-empty-details">{Object.keys(rootColumnSchemas).length}</span>
                )}
            />
        );
        expect(component.find(".custom-empty-details").text()).toBe("1");
    });
    it("with root array selection", () => {
        const columnData = [
            {
                items: {
                    "Schema One": {
                        title: "Schema Title",
                        items: { $ref: "itemSchema" }
                    }
                },
                selectedItem: "Schema One",
                trailingSelection: true
            }
        ];
        const refTargets = {
            itemSchema: {
                description: "Array Item",
                type: "object"
            }
        };
        const component = shallow(
            <InspectorDetails
                columnData={columnData}
                refTargets={refTargets}
            />
        );
        const forms = component.find("InspectorDetailsForm");
        expect(forms).toHaveLength(2);
        expect(forms.first().prop("fields")).toEqual([
            {
                labelText: "Title",
                rowValue: "Schema Title"
            }
        ]);
        expect(forms.last().prop("fields")).toEqual([
            {
                labelText: "Description",
                rowValue: "Array Item"
            },
            {
                labelText: "Type",
                rowValue: "object"
            }
        ]);
        expect(component.exists(".jsonschema-inspector-details-separator")).toBe(true);
        const headers = component.find(".jsonschema-inspector-details-header");
        expect(headers).toHaveLength(2);
        expect(headers.first().text()).toEqual("Details");
        expect(headers.last().text()).toEqual("Array Entry Details");
    });
    it("with root array selection and custom renderSelectionDetails", () => {
        const mainSchema = {
            title: "Schema Title",
            items: { $ref: "itemSchema" }
        };
        const columnDataProp = [
            {
                items: {
                    "Schema One": mainSchema
                },
                selectedItem: "Schema One",
                trailingSelection: true
            }
        ];
        const refTargetsProp = {
            itemSchema: {
                description: "Array Item",
                type: "object"
            }
        };
        const renderSelectionDetails = jest.fn(() => (<span className="custom-selection-details" />));
        const component = shallow(
            <InspectorDetails
                columnData={columnDataProp}
                refTargets={refTargetsProp}
                renderSelectionDetails={renderSelectionDetails}
            />
        );
        expect(component.exists(".custom-selection-details")).toBe(true);
        expect(renderSelectionDetails.mock.calls).toHaveLength(1);
        const {
            itemSchema, refTargets, columnData, selectionColumnIndex
        } = renderSelectionDetails.mock.calls[0][0];
        expect(itemSchema).toEqual(mainSchema);
        expect(refTargets).toEqual(refTargetsProp);
        expect(columnData).toEqual(columnDataProp);
        expect(selectionColumnIndex).toBe(0);
    });
});
describe("collectFormFields()", () => {
    it("includes `title`", () => {
        const schema = { title: "Title Value" };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Title",
                rowValue: "Title Value"
            }
        ]);
    });
    it("includes `title` from $ref-erenced schema", () => {
        const schema = { $ref: "A" };
        const refTargets = {
            A: { title: "Title Value" }
        };
        expect(InspectorDetails.collectFormFields(schema, refTargets, [{}], 0)).toEqual([
            {
                labelText: "Title",
                rowValue: "Title Value"
            }
        ]);
    });
    it("includes `description`", () => {
        const schema = { description: "Description Value" };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Description",
                rowValue: "Description Value"
            }
        ]);
    });
    it("includes `required`", () => {
        const schemaProperties = {
            field: { title: "Required Property" }
        };
        const columnData = [
            {
                items: {
                    "Schema One": {
                        required: ["field"],
                        properties: schemaProperties
                    }
                },
                selectedItem: "Schema One",
                trailingSelection: false
            },
            {
                items: schemaProperties,
                selectedItem: "field",
                trailingSelection: true
            }
        ];
        expect(InspectorDetails.collectFormFields({}, {}, columnData, 1)).toEqual([
            {
                labelText: "Required",
                rowValue: "Yes"
            }
        ]);
    });
    it("includes `type`", () => {
        const schema = { type: "object" };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Type",
                rowValue: "object"
            }
        ]);
    });
    it("includes `const`", () => {
        const schema = { const: 42 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Constant Value",
                rowValue: 42
            }
        ]);
    });
    it("includes `enum`", () => {
        const schema = { enum: [42, 84] };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Possible Values",
                rowValue: [42, 84]
            }
        ]);
    });
    it("includes `minimum` without exclusiveMinimum", () => {
        const schema = { minimum: 42 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Min Value",
                rowValue: "42 (inclusive)"
            }
        ]);
    });
    it("includes `minimum` with boolean exclusiveMinimum (as per Draft 4)", () => {
        const schema = {
            minimum: 42,
            exclusiveMinimum: true
        };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Min Value",
                rowValue: "42 (exclusive)"
            }
        ]);
    });
    it("includes `exclusiveMinimum` (as per Draft 6)", () => {
        const schema = {
            exclusiveMinimum: 42
        };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Min Value",
                rowValue: "42 (exclusive)"
            }
        ]);
    });
    it("includes `maximum` without `exclusiveMaximum`", () => {
        const schema = { maximum: 84 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Max Value",
                rowValue: "84 (inclusive)"
            }
        ]);
    });
    it("includes `maximum` with boolean `exclusiveMaximum` (as per Draft 4)", () => {
        const schema = {
            maximum: 84,
            exclusiveMaximum: true
        };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Max Value",
                rowValue: "84 (exclusive)"
            }
        ]);
    });
    it("includes `exclusiveMaximum` (as per Draft 6)", () => {
        const schema = {
            exclusiveMaximum: 84
        };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Max Value",
                rowValue: "84 (exclusive)"
            }
        ]);
    });
    it("includes `default` (object)", () => {
        const schema = { default: {} };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Default Value",
                rowValue: "{}"
            }
        ]);
    });
    it("includes `default` (non-object)", () => {
        const schema = { default: false };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Default Value",
                rowValue: false
            }
        ]);
    });
    it("includes `examples` (objects)", () => {
        const schema = {
            examples: [
                {
                    field: "value"
                }
            ]
        };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Example(s)",
                rowValue: "[{\"field\":\"value\"}]"
            }
        ]);
    });
    it("includes `examples` (non-objects)", () => {
        const schema = { examples: [42, 84] };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Example(s)",
                rowValue: [42, 84]
            }
        ]);
    });
    it("ignores empty `examples`", () => {
        const schema = { examples: [] };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([]);
    });
    it("includes `pattern`", () => {
        const schema = { pattern: "[a-z]+" };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Value Pattern",
                rowValue: "[a-z]+"
            }
        ]);
    });
    it("includes `format`", () => {
        const schema = { format: "iri" };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Value Format",
                rowValue: "iri"
            }
        ]);
    });
    it("includes `minLength`", () => {
        const schema = { minLength: 1 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Min Length",
                rowValue: 1
            }
        ]);
    });
    it("includes `maxLength`", () => {
        const schema = { maxLength: 100 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Max Length",
                rowValue: 100
            }
        ]);
    });
    it("includes `minItems`", () => {
        const schema = { minItems: 1 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Min Items",
                rowValue: 1
            }
        ]);
    });
    it("includes `maxItems`", () => {
        const schema = { maxItems: 100 };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Max Items",
                rowValue: 100
            }
        ]);
    });
    it("includes `uniqueItems`", () => {
        const schema = { uniqueItems: true };
        expect(InspectorDetails.collectFormFields(schema, {}, [{}], 0)).toEqual([
            {
                labelText: "Items Unique",
                rowValue: "Yes"
            }
        ]);
    });
});
