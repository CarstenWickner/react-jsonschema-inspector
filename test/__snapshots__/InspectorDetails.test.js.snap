// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-details"
>
  <InspectorDetailsContent
    columnData={
      Array [
        Object {
          "items": Object {
            "Schema One": JsonSchema {
              "getFieldValue": [Function],
              "getProperties": [Function],
              "getPropertyParentSchemas": [Function],
              "getSchemaFieldValue": [Function],
              "getTypeOfArrayItems": [Function],
              "schema": Object {
                "description": "Text",
                "title": "Schema Title",
              },
              "scope": RefScope {
                "addOtherScope": [Function],
                "addOtherScopes": [Function],
                "externalRefs": Map {},
                "find": [Function],
                "findSchemaInThisScope": [Function],
                "internalRefs": Map {
                  "#" => [Circular],
                },
                "otherScopes": Array [],
              },
            },
          },
          "selectedItem": "Schema One",
          "trailingSelection": true,
        },
      ]
    }
    itemSchema={
      JsonSchema {
        "getFieldValue": [Function],
        "getProperties": [Function],
        "getPropertyParentSchemas": [Function],
        "getSchemaFieldValue": [Function],
        "getTypeOfArrayItems": [Function],
        "schema": Object {
          "description": "Text",
          "title": "Schema Title",
        },
        "scope": RefScope {
          "addOtherScope": [Function],
          "addOtherScopes": [Function],
          "externalRefs": Map {},
          "find": [Function],
          "findSchemaInThisScope": [Function],
          "internalRefs": Map {
            "#" => [Circular],
          },
          "otherScopes": Array [],
        },
      }
    }
    selectionColumnIndex={0}
  />
</div>
`;
